README
NAME
    POE::Declare - A POE abstraction layer for conciseness and simplicity

SYNOPSIS
      package MyComponent;
      
  use strict;
      use POE::Declare {
          foo      => 'Attribute',
          bar      => 'Internal',
          Username => 'Param',
          Password => 'Param',
      };
      
  declare TimeoutError => 'Message';
      
  sub hello : Event {
          print "Hello World!\n";
          $_[SELF]->hello_timeout_start;
      }
      
  sub hello_timeout : Timeout(30) {
          print "Alas, I die!\n";
      
      # Tell our parent as well
          $_[SELF]->TimeoutError;
      }
      
  compile;

DESCRIPTION
    POE is a very powerful and flexible system for doing asynchronous
    programming. But it has the reputation of being difficult to learn, with
    a somewhat confusing set of abstractions.

    In particular, it can be tricky to resolve POE's way of programming with
    the highly abstracted OO structures that many people are used to, with
    layer stacked upon layer ad-infinitum to produce a delegation heirachy
    which allows for powerful and complex systems that are relatively easy
    to maintain.

    This can be particularly noticable as the scale of a POE codebase gets
    larger. At three levels of abstraction the layering in POE becomes quite
    difficult, and progess past about the third layer of abstraction becomes
    extremely difficult.

    POE::Declare is an attempt to resolve this problem by locking down part
    of the traditional flexibility of POE, and by making it easier to split
    the implementation of each object between an object-oriented heirachy
    and a collection of POE sessions.

    The goal is to retain the ability to build deep and complex heirachies
    of encapsulated functionality in your application while also allowing
    you to take advantage of the asynchronous nature of POE code.

  General Architecture
    At the core of any POE::Declare application is an object-oriented
    heirachy of components. This heirachy exists whether or not the POE
    kernel is running, and parts of it can be started and stopped as needed.

    When it is spawned, each component will create it's own private
    POE::Session in which to run its events and store its resources.

    Each instance of a class always has one and only one session. This may
    be a problem if your application will have thousands of spawned
    components at once (POE recommends against the use of large numbers of
    sessions) but should not be a problem as long as you keep it down to a
    few hundred components.

    Because the POE session is slaved to the POE::Declare::Object, the
    component can handle being started and stopped many times without the
    loss of any state data between the creation and destruction of each
    slave session.

    To allow support for components that have no resources and only act as
    supervisors, POE::Declare always assigns a POE alias for the session
    while it is active. The existance of this Alias prevents POE cleaning up
    the session accidentally, and ensures components have explicit control
    over when they want to shut down their sessions.

    Each POE component contains a set of named resources. Resources may
    consist of a different underlying POE resource, or may be made up of
    multiple resources, and additional data stored in the matching object
    HASH key. To ensure all the various underlying resources will not clash
    with each other, all resources must be declared and will be strictly
    checked.

    At the end of each class, instead of the usual 1; to allow the package
    to return true, you put instead a "compile;" statement.

    This instructs POE::Declare to inventory the declarations and
    attributes, combine them with declarations from the parent classes, and
    then generate the code that will implement the structures.

    Once the class has been compiled, the installed functions will be
    removed from the package to prevent run-time namespace pollution.

  Import-Time Declaration
    The cluster of "declare" statements at the beginning of a POE::Declare
    class can look ugly to some people, and may get annoying to step through
    in the debugger.

    To resolve this, you may optionally provide a list of slot declarations
    to the module at compile time. This should be in the form of a simple
    "HASH" reference with the names as keys and the type as values.

      use My::Module {
          Foo => 'Param',
          Bar => 'Param',
      };

    Event and timeout declarations cannot be provided by this method, and
    you should continue to use subroutine attributes for these as normal.

  Inheritance
    The resource model of POE::Declare correctly follows inheritance,
    similar to the way declarations in Moose are inherited. Resource types
    in a parent class cannot be overwritten or modified in child classes.

    No special syntax is needed for inheritance, as POE::Declare works
    directly on top of Perl's native inheritance.

      # Parent.pm - Object that connects to a service
      package Parent;
      
  use strict;
      use POE::Declare {
          Host           => 'Param',
          Port           => 'Param',
          ConnectSuccess => 'Message',
          ConnectFailure => 'Message',
      };
      
  sub connect : Event {
          # ...
      }
      
  compile;
      
  # Child.pm - Connect to an (optionally) authenticating service
      package Child;
      
  use strict;
      use base 'Parent';
      use POE::Declare {
          Username     => 'Param',
          Password     => 'Param',
          AuthRequired => 'Message',
          AuthInvalid  => 'Message',
      };
      
  compile;

CLASSES
    POE::Declare is composed of three main modules, and a tree of
    slot/attribute classes.

  POE::Declare
    POE::Declare provides the main interface layer and Domain Specific API
    for declaratively building your POE::Declare classes.

  POE::Declare::Object
    POE::Declare::Object is the abstract base class for all classes created
    by POE::Declare.

  POE::Declare::Meta
    POE::Declare::Meta implements the metadata structures that describe each
    of your POE::Declare classes. This is superficially similar to something
    like Moose, but unlike Moose is fast, light weight and can use
    domain-specific assumptions.

  POE::Declare::Slot
      POE::Declare::Meta::Slot
        POE::Declare::Meta::Internal
        POE::Declare::Meta::Attribute
          POE::Declare::Meta::Param
        POE::Declare::Meta::Message
        POE::Declare::Meta::Event
          POE::Declare::Meta::Timeout

  POE::Declare::Meta::Internal
    POE::Declare::Meta::Internal is a slot class that won't generate any
    functionality, but allows you to reserve an attribute for internal use
    so that they won't be used by any sub-classes.

  POE::Declare::Meta::Attribute
    POE::Declare::Meta::Attribute is a slot class used for readable
    attributes.

  POE::Declare::Meta::Param
    POE::Declare::Meta::Attribute is a slot class for attributes that are
    provided to the constructor as a parameter.

  POE::Declare::Meta::Message
    POE::Declare::Meta::Message is a slot class for declaring messages that
    the object will emit under various circumstances. Each message is a
    param to the constructor that takes a callback in a variety of formats
    (usually pointing up to the parent object).

  POE::Declare::Meta::Event
    POE::Declare::Meta::Event is a class for named POE events that can be
    called or yielded to by other POE messages/events.

  POE::Declare::Meta::Timeout
    POE::Declare::Meta::Timeout is a POE::Declare::Meta::Event sub-class
    that is designed to trigger from an alarm and generates additional
    methods to manage the alarms.

FUNCTIONS
    For the first few releases, I plan to leave this module undocumented.

    That I am releasing this distribution at all is more of a way to mark my
    progress, and to allow other POE/OO people to look at the implementation
    and comment.

  declare
      declare one   => 'Internal';
      declare two   => 'Attribute';
      declare three => 'Param';
      declare four  => 'Message';

    The "declare" function is exported by default. It takes two parameters,
    a slot name and a slot type.

    The slot name can be any legal Perl identifier.

    The slot type should be one of "Internal", "Attribute", "Param" or
    "Message".

    Creates the new slot, throws an exception on error.

  compile
    The "compile" function indicates that all attributes and events have
    been defined and the structure should be finalised and compiled.

    Returns true or throws an exception.

SUPPORT
    Bugs should be always be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Declare>

    For other issues, or commercial enhancement or support, contact the
    author.

AUTHORS
    Adam Kennedy <adamk@cpan.org>

SEE ALSO
    POE, <http://ali.as/>

COPYRIGHT
    Copyright 2006 - 2011 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Changes for Perl extension POE-Declare

0.58 Fri 10 Jun 2011
	- It is now legal to ->finish a session twice, as long as the session
	  still exists in POE terms. The second call will silently shortcut.

0.57 Thu 9 Jun 2011
	- Improved the error message from ->finish when the session does not
	  exist at the time that ->finish is called.

0.56 Tue 31 May 2011
	- Bumping POE and Test::POE::Stopping dependencies to get past the
	  POE::API::Peek explosion.

0.55 Fri 27 May 2011
	- Updating Module::Install::DSL to 1.01
	- Updating Test::POE::Stopping dependency to 1.08 to be able to deal
	  with the change to session internals in POE 1.310.
	- Test scripts compensate for a hardcoded assumption that the ID of the
	  first session is 2 in POE, it is now 1 (and the tests adapt)

0.54 Fri 15 Apr 2011
	- Add support for timeout variance both in declaration and overridden
	  timeout lengths.
	- Overriding timeout delays were being ignored (major bug fix)

0.53 Wed 9 Mar 2011
	- Fixed #51697: [Patch] POD nit (FWIE)
	- Upgrade Class::XSAccessor dependency to 1.10 to fix some crashes
	  that occur during END in certain circumstances. Use of POE::Wheel::Run
	  seems to be a particular culprit (maybe).

0.52 Mon 7 Mar 2011
	- Adding check for event declaration on finalised class
	- Adding check for use of an uncompiled class
	- Moved dependency version checks into module code
	- Removing apparently redundant IO::File dependency

0.51 Wed 23 Feb 2011
	- Renaming _compile to as_perl
	- Bumped the POE dependency again
	- Fixed a POD typo

0.50 Thu 21 Oct 2010
	- Major upgrade, not backward compatible in all cases.
	- Bumped the POE dependency as I am struggling to test against
	  older ones.
	- Timeout events are now implemented using POE named timers
	  rather than numbered timers. The implementation is much
	  closer to POE's prefered usage, and results in significantly
	  less code in 80% less code in generated timeout code.
	- Added the new method ->finish, which will clear all timers
	  and remove the session alias. It should be called when you
	  are ready for your session to be shut down.
	  The finish method MUST be called while in the session you
	  are calling it on. Removed the _finish event.
	- Removed the _alias_set and _alias_remove events. With the
	  introduction of the finish method, we fully control all
	  the places where the session alias needs to be created and
	  removed. Direct user access to this is now unneeded and
	  likely to be a source of bugs.
	- Any code previously using _alias_remove directly to do their
	  equivalent of what the finish method now does will break.

0.26 Mon 19 Jul 2010
	- Updating to Module::Install 1.00
	- Added _finish event which clears all timeouts and removes the
	  alias. Intended to be a common way to say "I'm done now".
	- Added alternate compile-time hash usage for declarations.
	- Corrected timeout_start bug where restarts were being done via
	  non-existant method timeout_clear.

0.25 Thu 27 May 2010
	- Updating to Module::Install 0.97
	- Updating Class::XSAccessor dependency to faster 1.05
	- Expanding documentation

0.24 Thu  8 Oct 2009
	- Moving child disconnection change to production version.
	- Bumping Test::POE::Stopping dependency to 1.05 for increased
	  rigour of stop validation.

0.23_01 Fri 11 Sep 2009
	- Experimental change, immediately disconnect every session from
	  their parent at _start time so that we never use POE's internal
	  child session reference tracking.

0.22 Wed 12 Aug 2009
	- Fixing some bugs in Message events, and improved tests for it.

0.21 Thu 19 Mar 2009
        - Improved checking of Message params
        - Adding an Alias format check

0.20 Tue 17 Mar 2009
	- Disable Test::NoWarnings when using Test::POE::Stopping.
	  It uses POE::Peek::API, which will throw a warnings when
	  loaded against a development version of POE.
	- Adding a specific META.yml dependency check.

0.19 Tue 17 Mar 2009
	- Allowing Message params to be optional

0.18 Fri 13 Mar 2009
	- Changing the Message implementation to be much simpler

0.17 Fri 13 Mar 2009
	- Adding POD documentation for _alias_remove
	- Moving the events into their own documentation section

0.16 Fri 13 Mar 2009
	- Adding missing Test::Exception dependency
	- Dropping the superfluous timeout_start method

0.15 Fri 13 Mar 2009
	- Moving package_states and params to private methods
	- Add strict checking of constructor params
	- POE::Declare::Object->meta is not code-generated per-class

0.14 Thu 12 Mar 2009
	- Adding the initial implementation of Timeout that actually works

0.13 Wed 11 Mar 2009
	- Adding a missing dependency on Test::NoWarnings

0.12 Tue 10 Mar 2009
	- Adding a missing dependency on Test::POE::Stopping

0.11 Mon  9 Mar 2009
	- Adding use warnings to all classes
	- Adding Test::NoWarnings to all tests
	- Fixing some bugs in _alias_remove

0.10 Mon  9 Mar 2009
	- Adding specific _alias_set and _alias_remove methods

0.09 Thu  5 Mar 2009
	- Converting all simple accessors to Class::XSAccessor
	- Switching to a better and faster error check in lookback

0.08 Thu  5 Mar 2009
	- Adding tests for half a dozen or so more methods
	- Fixed a bug in ->lookback

0.07 Thu  5 Mar 2009
	- compile is now prototyped ()
	- package_states caches now, making spawning faster

0.06 Wed  4 Mar 2009
	- use POE::Declare is now also an implicit use POE

0.05 Mon 23 Feb 2009
	- Completing most of the initial pass over the documentation
	- Switching to using Class::XSAccessor

0.04 skipped

0.03 Fri 20 Feb 2009
	- Upgrading to Module::Install 0.79
	- Correcting dependencies
	- Timeout events are modelled correctly now
	- No functional changes

0.02 Tue 12 Sep 2006
	- More thorough testing
	- Fixing various bugs
	- Added event auto-registration
	- Added a base meta-model for POE::Declare::Object, removing the special case

0.01 Thu  7 Sep 2006
	- original version


inc/Module/Install/With.pm
#line 1
package Module::Install::With;

# See POD at end for docs

use strict;
use Module::Install::Base;

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.80';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}





#####################################################################
# Installer Target

# Are we targeting ExtUtils::MakeMaker (running as Makefile.PL)
sub eumm {
	!! ($0 =~ /Makefile.PL$/i);
}

# You should not be using this, but we'll keep the hook anyways
sub mb {
	!! ($0 =~ /Build.PL$/i);
}





#####################################################################
# Testing and Configuration Contexts

#line 49

sub interactive {
	# Treat things interactively ONLY based on input
	!! (-t STDIN and ! automated_testing());
}

#line 67

sub automated_testing {
	!! $ENV{AUTOMATED_TESTING};
}

#line 86

sub release_testing {
	!! $ENV{RELEASE_TESTING};
}

sub author_context {
	!! $Module::Install::AUTHOR;
}





#####################################################################
# Operating System Convenience

#line 114

sub win32 {
	!! ($^O eq 'MSWin32');
}

#line 131

sub winlike {
	!! ($^O eq 'MSWin32' or $^O eq 'cygwin');
}

1;

#line 159


lib/POE/Declare/Meta/Timeout.pm
package POE::Declare::Meta::Timeout;

=pod

=head1 NAME

POE::Declare::Meta::Timeout - A named timeout with generated support methods

=head1 SYNOPSIS

  # Send a request
  sub request : Event {
      $_[SELF]->request_timeout_start;
      $_[SELF]->{handle}->put('something');
  }
  
  # Received a response
  sub response : Event {
      if ( $_[ARG0] eq 'keepalive' ) {
          $_[SELF]->request_timeout_restart;
          return;
      }
  
      $_[SELF]->request_timeout_stop;
      $_[SELF]->{parent}->post('child_response', $_[ARG0]);
  }
  
  # Did not get a response
  sub request_timeout : Timeout(30) {
      # Take some action
      $_[SELF]->cleanup;
      $_[SELF]->{parent}->post('child_response', 'timeout');
  }

=head1 DESCRIPTION

B<POE::Declare::Meta::Timeout> is a sub-class of C<Event> with access to
a number of additional methods relating to timers and alarms.

=cut

use 5.008007;
use strict;
use warnings;
use POE::Declare::Meta::Event ();

use vars qw{$VERSION @ISA};
BEGIN {
	$VERSION = '0.58';
	@ISA     = 'POE::Declare::Meta::Event';
}

use Class::XSAccessor {
	getters => {
		delay    => 'delay',
		variance => 'variance',
	},
};





#####################################################################
# Main Methods

sub as_perl {
	my $name     = $_[0]->{name};
	my $delay    = $_[0]->{delay};
	my $variance = $_[0]->{variance} || 0;
	return <<"END_PERL";
sub ${name}_start {
	my \$delay    = \@_ > 1 ? defined \$_[1] ? \$_[1] : $delay : $delay;
	my \$variance = \@_ > 1 ? \$_[2] || $variance : $variance;
	\$poe_kernel->delay(
		'$name' => \$delay + rand(\$variance * 2) - \$variance
	) or return;
	die("Failed to create timeout for event '$name'");
}

*${name}_restart = *${name}_start;

sub ${name}_stop {
	my \$self  = (\@_ == 1) ? \$_[0] : \$_[HEAP];
	\$poe_kernel->delay( '$name' ) or return;
	die("Failed to create timeout for event '$name'");
}
END_PERL
}

1;

=pod

=head1 SUPPORT

Bugs should be always be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Declare>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHORS

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 SEE ALSO

L<POE>, L<POE::Declare>

=head1 COPYRIGHT

Copyright 2006 - 2011 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
